src/main/java/org/magicdgs/readtools/utils/read/stats/engine/ProperStatWindowEngine.java

```diff
@@ -0,0 +1,275 @@
+import java.io.PrintStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Engine to compute several {@link StatFunction} over the provided windows.
+ *
+ * <p>This class takes a list of intervals and {@link StatFunction} and expects reads to be added
+ * to compute on-the-fly the statistics provided. Nevertheless, the engine has the following
+ * assumptions that the caller of {@link #addRead(GATKRead)} should met:
+ *
+ * <p>The engine has the following assumptions to work:
+ *
+ * <ul>
+ *     <li>Provided intervals are coordinate-sorted.</li>
+ *     <li>Reads are added in a coordinate-sorted manner.</li>
+ *     <li>Only two fragments per template will be added (supplementary/secondary alignments are filtered.</li>
+ * </ul>
+ *
+ * <p>If these assumptions are not met, windows are output without all the data and statistics
+ * would be incorrect.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class ProperStatWindowEngine implements Closeable {
+
+    private final Logger logger = LogManager.getLogger(this.getClass());
+
+    // list with all output windows
+    // TODO: possible performance improvements:
+    // TODO: - use IntervalsSkipList: retrieve overlapping intervals per read and add only there
+    // TODO: - it also requires to retrieve overlapping windows for the mate (extract location)
+    private final SortedMap<String, List<ProperStatWindowCalculator>> windowsPerContig;
+
+    // cache column names
+    private final List<String> columnNames;
+    // option to print all the windows
+    private final boolean printAll;
+    // ouptut writer
+    private final PrintStream writer;
+
+    /**
+     * Constructor for the engine.
+     *
+     * @param dict sequence dictionary to use with intervals.
+     * @param intervals coordinate-sorted intervals to compute the statistics.
+     * @param singleStats statistics for single-reads (without considering pairs).
+     * @param pairStats statistics for pair-reads (considering the mate).
+     * @param output output file to populate the results.
+     * @param printAll if {@code false}, windows with 0 reads will be discarded; otherwise,
+     *                 they will be populated.
+     * @throws IOException if an I/O error occurs while opening the output.
+     */
+    public ProperStatWindowEngine(final SAMSequenceDictionary dict,
+            final List<SimpleInterval> intervals,
+            final List<SingleReadStatFunction> singleStats,
+            final List<PairEndReadStatFunction> pairStats,
+            final Path output,
+            final boolean printAll) throws IOException {
+        // TODO: maybe support also compressed output if end with a compressed extension?
```

Comment:
Open an issue for this.

Reply:
Done and complete with other improvements (commented on the `PrintStream` instead.